### PR TITLE
incorporate csawMG update and pass both intel and GNU RT tests

### PR DESCRIPTION
**Pull request by grantfirl**
This replaces https://github.com/ufs-community/ccpp-physics/pull/171. This version started from the top of ufs/dev and only added necessary changes. I've verified that this gives bit-for-bit answers compared to https://github.com/ufs-community/ccpp-physics/pull/171 in the SCM.

Description from @AnningCheng-NOAA:

Major updated CSAW deep convective scheme: considering sigma vertical veriantion and reduction on fluxes instead of tendency etc.
• Regression tests are performed on the new version of CSAW for GNU and intel compilers
Here is a EMail from Don describe the changes:
On Thu, Oct 24, 2019 at 3:01 PM Donald Dazlich - NOAA Affiliate <[donald.dazlich@noaa.gov](mailto:donald.dazlich@noaa.gov)> wrote:
Hi Moorthi, Anning,
Dave and I have been using the GFS version of CS to redo our implementation of AW more completely and consistently. Now, you will find that the fix_form implemetation of the tendencies match up very closely with the traditional form. Also you will notice that it is the fluxes for each cloud type (and the downdraft) that are saved and the reduction is done on the fluxes, not the tendency computed from the flux divergence. Since sigma is a function of height the latter approach is incorrect. I’ve also dimensioned several cloud type variables nctp+1 - the nctp+1 slot has downdraft output. Another addition is the sorting of the lamdas after they are computed for all cloud types, then computing the sigma. This approach is detailed in a manuscript Dave is working on. Also, sigma is not applied to the inputs to the downdraft routine, but to the outputs, which are now mass fluxes not tendencies. The water and thermodynamic budgets remain satisfied now.
I’ve attached my version below. It should be pretty close to your version with regards to the interface so you shouldn’t have to do much work with it.
Regards,
Don

### CODE REVIEW

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (grantfirl) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:3852]: @AnningCheng-NOAA The code in https://github.com/ufs-community/ccpp-physics/pull/171 wouldn't compile without adding the `NTR` dimension back to `GTR` since the third dimension of `GTR` is used in this subroutine. Please double-check that this is OK. This is the only **CODE** change in https://github.com/ufs-community/ccpp-physics/pull/171 that was reverted.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (grantfirl) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:507]: @AnningCheng-NOAA I'm guessing that this doesn't get executed since `do_aw` is true when this scheme is executed in UFS, but if this branch of code does get executed, the calculation of `w_upi` is referencing `cf_upi` that no longer is given a value.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (grantfirl) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90]: @AnningCheng-NOAA One of the CCPP rules is that all intent(OUT) variables are given values. These comments prevent `cf_upi` from being given a value. If cf_upi is no longer needed, we can remove it from the argument list and CCPP metadata, although see the comment below about the calculation of `w_upi` when `do_aw = F`.

**Comment by grantfirl** [comment]
@dustinswales @Qingfu-Liu Please review this again when you get a minute. It should be easier to follow now.

**Comment by AnningCheng-NOAA** [comment]
@Grant Firl ***@***.***>  please add back.

**Comment by AnningCheng-NOAA** [comment]
Either way is fine with me. Those variables are for MG usage and have been
tuned for best performance. It seems like w_upi=0 and not using cf_upi had
been chosen although I tend to keep the code  by uncommenting cf_upi in
case for future reference.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (AnningCheng-NOAA) [reply physics/CONV/Chikira_Sugiyama/cs_conv.F90:507]: I prefer to give a value to cf_up by uncommenting line 495.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (AnningCheng-NOAA) [reply physics/CONV/Chikira_Sugiyama/cs_conv.F90]: Please keep line 469 and uncomment it. see my comment below.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (AnningCheng-NOAA) [reply physics/CONV/Chikira_Sugiyama/cs_conv.F90:3852]: yes, PLease adding NTR back.

**Review by AnningCheng-NOAA — APPROVED** [review]
The code is fine with me now.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:1277]: Lines 1245-1277 can be removed

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:1796]: lines 1779-1796 can be removed

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:3330]: lines 3315-3330 can be removed

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:1589]: lines 1586-1589 can be removed

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:1142]: lines 1132-1142 can be removed.

File: physics/CONV/Chikira_Sugiyama/cs_conv.F90
Comment (Qingfu-Liu) [inline physics/CONV/Chikira_Sugiyama/cs_conv.F90:1809]: lines 1806-1809 can be removed

**Comment by zach1221** [comment]
@Qingfu-Liu @dustinswales @mzhangw regression testing is complete on UFS-WM PR [#2180](https://github.com/ufs-community/ufs-weather-model/pull/2180) , and ready for final review.

**Comment by grantfirl** [comment]
@Qingfu-Liu Please let me know if you feel strongly about removing the commented lines in this PR and I'll do that before merge.

**Review by dustinswales — APPROVED** [review]
These changes look great to me. Thanks for all the good work!

**Comment by zach1221** [comment]
Ok, I think we have enough approvals to merge at this time. @grantfirl

**Pull request merged**
